hi-ml-histopathology/src/histopathology/utils/k_top_bottom_tiles_utils.py

```diff
@@ -0,0 +1,403 @@
+
+        _, bottom_k_indices = torch.topk(attn_scores.squeeze(), k=k_tiles, largest=False, sorted=True)
+        self.bottom_tiles = [TileNode(data=tiles[i], attn=attn_scores[i].item()) for i in bottom_k_indices]
+
+    def shallow_copy(self) -> "SlideNode":
+        """Returns a shallow copy of the current slide node contaning only the slide_id and its probability score."""
+        return SlideNode(self.slide_id, self.prob_score)
+
+    def plot_attention_tiles(
+        self, top: bool, case: str, key_dir: Path, ncols: int = 5, size: Tuple[int, int] = (10, 10)
+    ) -> None:
+        """Plot and dave top or bottom tiles figures with their attention scores.
+
+        :param top: A boolean flag to pick top or bottom tiles.
+        :param case: The report case (e.g., TP, FN, ...)
+        :param key_dir: The path to the directory where to save the attention tiles figure.
+        :param ncols: Number of columns to create the subfigures grid, defaults to 5
+        :param size: The figure size , defaults to (10, 10)
+        """
+        tiles = self.top_tiles if top else self.bottom_tiles
+        suffix = "top" if top else "bottom"
+        nrows = int(ceil(len(tiles) / ncols))
+        if nrows > 0:
+            fig, axs = plt.subplots(nrows=nrows, ncols=ncols, figsize=size)
+            fig.suptitle(f"{case}: {self.slide_id} P=%.2f" % abs(self.prob_score))
+
+            for i, tile_node in enumerate(tiles):
+                axs.ravel()[i].imshow(np.transpose(tile_node.data.cpu().numpy(), (1, 2, 0)), clim=(0, 255), cmap="gray")
+                axs.ravel()[i].set_title("%.6f" % tile_node.attn)
+
+            for i in range(len(axs.ravel())):
+                axs.ravel()[i].set_axis_off()
+
+            save_figure(fig=fig, figpath=key_dir / f"{self.slide_id}_{suffix}.png")
+
+
+SlideOrTileKey = Union[SlideKey, TileKey]
+SlideDict = Dict[int, List[SlideNode]]
+TileDict = Dict[str, List[TileNode]]
+SlideOrTileDict = Union[SlideDict, TileDict]
+
+
+class KTopBottomTilesHandler:
+    """Class that manages selecting top and bottom tiles on the fly during validation and test of DeepMIL models."""
+
+    def __init__(self, n_classes: int, k_slides: int = 10, k_tiles: int = 10, ncols: int = 4) -> None:
+        """
+        :param n_classes: Number of MIL classes (set `n_classes=1` for binary).
+        :param k_slides: Number of slides to select to define top and bottom tiles based of pred scores. Defaults to 10.
+        :param k_tiles: Number of tiles to select as top and bottom tiles based on attn scores. Defaults to 10.
+        :param ncols: Number of columns to use to plot top and bottom tiles.
+        """
+        self.n_classes = n_classes if n_classes > 1 else 2
+        self.k_slides = k_slides
+        self.k_tiles = k_tiles
+        self.ncols = ncols
+        self.top_slides_heaps: SlideDict = {class_id: [] for class_id in range(self.n_classes)}
```

Comment:
you're using `Dict[int, List[SlideNode]]` a LOT, maybe define a type abbreviation for consistency?